File: tp_3_design_pattern/Python/modeles/etudiant.py
# etudiant.py - the Student class
# it inherits from Personne and uses Observer, Decorator and Strategy patterns

from __future__ import annotations

from typing import List

from modeles.personne import Personne
from modeles.cours import Cours
from patterns.observer import ISubject, IObserver
from patterns.decorator import IEtudiantComponent
from patterns.strategy import ICalculMention, MentionStandard


class Etudiant(Personne, ISubject, IEtudiantComponent):
    """Represents a student in the academic system."""

    def __init__(
        self,
        nom: str,
        age: int,
        numero_etudiant: str,
        moyenne: float = 0.0,
        strategie_mention: ICalculMention | None = None,
    ):
        super().__init__(nom, age)
        self.numero_etudiant = numero_etudiant
        self._moyenne = moyenne
        self._cours: List[Cours] = []
        self._notes: List[float] = []
        self._observers: List[IObserver] = []
        # if no strategy is given, we use the standard one
        self._strategie_mention = strategie_mention or MentionStandard()

    @property
    def moyenne(self) -> float:
        return self._moyenne

    @property
    def cours(self) -> List[Cours]:
        return self._cours

    def ajouter_cours(self, cours: Cours) -> None:
        self._cours.append(cours)

    # Strategy pattern - we can change the mention calculation at runtime
    def set_strategie_mention(self, strategie: ICalculMention) -> None:
        self._strategie_mention = strategie

    def calculer_mention(self) -> str:
        return self._strategie_mention.calculer(self._moyenne)

    def ajouter_note(self, note: float) -> None:
        if note < 0 or note > 20:
            raise ValueError("La note doit être comprise entre 0 et 20.")
        self._notes.append(note)
        self._moyenne = sum(self._notes) / len(self._notes)
        # notify all observers that a new grade was added
        self.notifier("note_ajoutee")

    def afficher(self) -> str:
        noms_cours = (
            ", ".join(c.nom_cours for c in self._cours)
            if self._cours
            else "Aucun cours"
        )
        return (
            f"Etudiant(nom={self.nom}, age={self.age}, num={self.numero_etudiant}, "
            f"moyenne={self.moyenne:.2f}, mention={self.calculer_mention()}, "
            f"cours=[{noms_cours}])"
        )

    # Observer pattern methods
    def attacher(self, observer: IObserver) -> None:
        if observer not in self._observers:
            self._observers.append(observer)

    def detacher(self, observer: IObserver) -> None:
        if observer in self._observers:
            self._observers.remove(observer)

    def notifier(self, evenement: str) -> None:
        for obs in self._observers:
            obs.update(self, evenement)

File: tp_3_design_pattern/Java/modeles/Etudiant.java
package modeles;

import java.util.ArrayList;
import java.util.List;

import patterns.observer.IObserver;
import patterns.observer.ISubject;
import patterns.strategy.ICalculMention;
import patterns.strategy.MentionStandard;
import patterns.decorator.IEtudiantComponent;

// the student class - inherits from Personne
// implements ISubject (Observer pattern) so it can notify when grades are added
// implements IEtudiantComponent (Decorator pattern) so it can be decorated
public class Etudiant extends Personne implements ISubject, IEtudiantComponent {

    private String numeroEtudiant;
    private double moyenne;
    private List<Cours> cours;
    private List<Double> notes;
    private List<IObserver> observers;
    private ICalculMention strategieMention;

    public Etudiant(String nom, int age, String numeroEtudiant,
                    double moyenne, ICalculMention strategieMention) {
        super(nom, age);
        this.numeroEtudiant = numeroEtudiant;
        this.moyenne = moyenne;
        this.strategieMention = (strategieMention != null)
                ? strategieMention
                : new MentionStandard();
        this.cours = new ArrayList<>();
        this.notes = new ArrayList<>();
        this.observers = new ArrayList<>();
    }

    public String getNumeroEtudiant() {
        return numeroEtudiant;
    }

    public double getMoyenne() {
        return moyenne;
    }

    public List<Cours> getCours() {
        return cours;
    }

    public void ajouterCours(Cours cours) {
        this.cours.add(cours);
    }

    // Strategy pattern - change the mention calculation at runtime
    public void setStrategieMention(ICalculMention strategieMention) {
        this.strategieMention = strategieMention;
    }

    public String calculerMention() {
        return strategieMention.calculer(moyenne);
    }

    // when a grade is added, we recalculate the average and notify observers
    public void ajouterNote(double note) {
        if (note < 0 || note > 20) {
            throw new IllegalArgumentException(
                    "La note doit être comprise entre 0 et 20.");
        }
        notes.add(note);

        double somme = 0;
        for (double n : notes) {
            somme += n;
        }
        moyenne = somme / notes.size();

        notifier("note_ajoutee");
    }

    @Override
    public String afficher() {
        StringBuilder nomsCours = new StringBuilder();
        if (cours.isEmpty()) {
            nomsCours.append("Aucun cours");
        } else {
            for (int i = 0; i < cours.size(); i++) {
                nomsCours.append(cours.get(i).getNomCours());
                if (i < cours.size() - 1) {
                    nomsCours.append(", ");
                }
            }
        }

        return "Etudiant(nom=" + nom
                + ", age=" + age
                + ", num=" + numeroEtudiant
                + ", moyenne=" + String.format("%.2f", moyenne)
                + ", mention=" + calculerMention()
                + ", cours=[" + nomsCours + "])";
    }

    // Observer pattern methods
    @Override
    public void attacher(IObserver observer) {
        if (!observers.contains(observer)) {
            observers.add(observer);
        }
    }

    @Override
    public void detacher(IObserver observer) {
        observers.remove(observer);
    }

    @Override
    public void notifier(String evenement) {
        for (IObserver obs : observers) {
            obs.update(this, evenement);
        }
    }
}
